ssl/quic/quic_srtm.c

```diff
@@ -0,0 +1,518 @@
+}
+
+QUIC_SRTM *ossl_quic_srtm_new(OSSL_LIB_CTX *libctx, const char *propq)
+{
+    QUIC_SRTM *srtm = NULL;
+    OSSL_PARAM params[3], *p = params;
+    unsigned char key[16];
+
+    if ((srtm = OPENSSL_zalloc(sizeof(*srtm))) == NULL)
+        return NULL;
+
+    /*
+     * Construct our blinding context using a random key. Since this only exists
+     * for side channel mitigation, we don't need to make the key configurable
+     * and simply use a random key each time. We never reconfigure the
+     * EVP_MAC_CTX after setup so we don't need to keep the key around.
+     */
+    if ((srtm->blind_mac = EVP_MAC_fetch(libctx, "CMAC", propq)) == NULL)
+        goto err;
+
+    if ((srtm->blind_ctx = EVP_MAC_CTX_new(srtm->blind_mac)) == NULL)
+        goto err;
+
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_CIPHER,
+                                            "AES-128-CBC", 12);
+    if (propq != NULL)
+        *p++ = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_PROPERTIES,
+                                                (char *)propq, 0);
+
+    *p++ = OSSL_PARAM_construct_end();
+
+    if (RAND_bytes_ex(libctx, key, sizeof(key), sizeof(key) * 8) != 1)
+        goto err;
+
+    if (!EVP_MAC_init(srtm->blind_ctx, key, sizeof(key), params))
+        goto err;
+
+    /* Create mappings. */
+    if ((srtm->items_fwd = lh_SRTM_ITEM_new(items_fwd_hash, items_fwd_cmp)) == NULL
+        || (srtm->items_rev = lh_SRTM_ITEM_new(items_rev_hash, items_rev_cmp)) == NULL)
+        goto err;
+
+    return srtm;
+
+err:
+    /*
+     * No cleansing of key needed as blinding exists only for side channel
+     * mitigation.
+     */
+    ossl_quic_srtm_free(srtm);
+    return NULL;
+}
+
+static void srtm_free_each(SRTM_ITEM *ihead)
+{
+    SRTM_ITEM *inext, *item = ihead;
+
+    for (item = item->next_by_seq_num; item != NULL; item = inext) {
+        inext = item->next_by_seq_num;
+        OPENSSL_free(item);
+    }
+
+    OPENSSL_free(ihead);
+}
+
+void ossl_quic_srtm_free(QUIC_SRTM *srtm)
+{
+    if (srtm == NULL)
+        return;
+
+    lh_SRTM_ITEM_free(srtm->items_rev);
+    if (srtm->items_fwd != NULL) {
+        lh_SRTM_ITEM_doall(srtm->items_fwd, srtm_free_each);
+        lh_SRTM_ITEM_free(srtm->items_fwd);
+    }
+
+    EVP_MAC_CTX_free(srtm->blind_ctx);
+    EVP_MAC_free(srtm->blind_mac);
+    OPENSSL_free(srtm);
+}
+
+/*
+ * Find a SRTM_ITEM by (opaque, seq_num). Returns NULL if no match.
+ * If head is non-NULL, writes the head of the relevant opaque list to *head if
+ * there is one.
+ * If prev is non-NULL, writes the previous node to *prev or NULL if it is
+ * the first item.
+ */
+static SRTM_ITEM *srtm_find(QUIC_SRTM *srtm, void *opaque, uint64_t seq_num,
+                            SRTM_ITEM **head_p, SRTM_ITEM **prev_p)
+{
+    SRTM_ITEM key, *item = NULL, *prev = NULL;
+
+    key.opaque  = opaque;
+
+    item = lh_SRTM_ITEM_retrieve(srtm->items_fwd, &key);
```

Comment:
This is a pattern I see alot, in which we retrieve an item from a hash table and return it to a caller.  Clearly it generally works, as asan doesn't detect leaks all over the place, but shouldn't we be doing some refcounting here to ensure that returned items aren't being referenced by callers when we evenually free the hash table?  Or is it a point of the design that all users of this api will have shut down prior to the free operation?